### PR TITLE
sql: enable clearing user password

### LINKED ISSUE #43891: sql: it's not possible to clear a user's password
- today it's not possible to remove a user's password after it was set
- it's useful to clear a user's password because this is a sure way to enforce they authenticate using client certificates
- pg allows this with `ALTER USER [WITH] PASSWORD NULL`

This should be supported in crdb as well.

### PR DESCRIPTION
**Pull request by knz**
Fixes #43891. Needed/useful for #43847.

It is useful to clear a user's password, as this is the only way to
enforce they authenticate using client certificates regardless of the
HBA configuration.

Prior to this patch, it was not possible to clear a user's password.
This patch makes it possible.

Release note (security update): Previously if a user password's had
been set once, thus enabling them to use password authentication, it
was not possible revert that choice and disable password
authentication for them any more other than either dropping the user,
or adding a specific per-user HBA rule. This has been fixed and the
PostgreSQL-compatible `ALTER USER WITH PASSWORD NULL` statement can
now be used to clear the user's password.

Release note (sql change): CREATE USER and ALTER USER now accept the
parameter `[WITH] PASSWORD NULL` to indicate the user's password
should be removed, thus preventing them from using password
authentication. This is compatible with PostgreSQL.

### CODE REVIEW

File: pkg/sql/pgwire/testdata/auth/password_change [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+# Erasing the password forces cert authentication.
+
+sql
+ALTER USER userpw WITH PASSWORD NULL
+----
+ok
+
+connect user=userpw password=pass2
```
Comment (mjibson) [inline pkg/sql/pgwire/testdata/auth/password_change]: Would a `connect user=userpw` test here also be useful?

File: pkg/sql/pgwire/testdata/auth/password_change [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+----
+ok defaultdb
+
+
+# Erasing the password forces cert authentication.
+
+sql
+ALTER USER userpw WITH PASSWORD NULL
```
Comment (mjibson) [inline pkg/sql/pgwire/testdata/auth/password_change]: do we need to test that you cannot set an empty password?

**Review by madelynnblue — APPROVED** [review]

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43892)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson)

___
*[pkg/sql/pgwire/testdata/auth/password_change, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43892#-LyJlZzb5j0jwRqanEK0-r1-42:-LyYrhyGAJZrX2uqAV_I:b-7i7nn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b9f0bda546fd376f76bb8e5c2b322a5451d2f0/pkg/sql/pgwire/testdata/auth/password_change#L42)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

do we need to test that you cannot set an empty password?
</blockquote></details>

There are such tests in the logictest module already (`testdata/logic_test/user`).
___
*[pkg/sql/pgwire/testdata/auth/password_change, line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43892#-LyJlZzb5j0jwRqanEK0-r1-46:-LyYs61D6Bepgc8TNOrO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8b9f0bda546fd376f76bb8e5c2b322a5451d2f0/pkg/sql/pgwire/testdata/auth/password_change#L46)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Would a `connect user=userpw` test here also be useful?
</blockquote></details>

Done.

**Comment by knz** [comment]
TFYR!

bors r+

**Pull request merged**